src/v/archival/archiver_service.cc

```diff
@@ -0,0 +1,999 @@
+      , _self_id(broker_id)
+      , _part(std::move(part))
+      , _config(std::move(config))
+      , _remote(remote)
+      , _cache(cache)
+      , _upload_housekeeping(housekeeping)
+      , _rtc(_as)
+      , _ctxlog(
+          archival_log,
+          _rtc,
+          ssx::sformat("{} node-{}", _ntp.path(), _self_id)) {
+        vlog(_ctxlog.debug, "created disposing managed_partition");
+    }
+
+    ~managed_partition_stm() {
+        vlog(_ctxlog.debug, "disposing managed_partition_stm");
+        vassert(_archiver == nullptr, "archiver is not stopped {}", _ntp);
+    }
+
+    managed_partition_stm() = delete;
+    managed_partition_stm(const managed_partition_stm&) = delete;
+    managed_partition_stm(managed_partition_stm&&) noexcept = delete;
+    managed_partition_stm& operator=(const managed_partition_stm&) = delete;
+    managed_partition_stm& operator=(managed_partition_stm&&) noexcept = delete;
+
+private:
+    /// Returns true if the data should be uploaded to S3 which means that
+    /// either topic config or local config is set and read replica mode is
+    /// disabled. If legacy mode is enabled always return false.
+    bool should_construct_archiver() const {
+        const auto& ntp_config = _part->log()->config();
+        return config::shard_local_cfg().cloud_storage_enabled()
+               && !config::shard_local_cfg()
+                     .cloud_storage_enable_legacy_upload_mode()
+               && (ntp_config.is_archival_enabled() || ntp_config.is_read_replica_mode_enabled());
+    }
+
+    /// Construct and return archiver but only if 'should_construct_archiver'
+    /// returns 'true'
+    ss::lw_shared_ptr<ntp_archiver> maybe_construct_archiver() const {
+        if (should_construct_archiver()) {
+            const auto& ntp_config = _part->log()->config();
+            auto manifest_view = _part->get_cloud_storage_manifest_view();
+            auto archiver = ss::make_lw_shared<archival::ntp_archiver>(
+              ntp_config,
+              _config,
+              _remote,
+              _cache,
+              *_part.get(),
+              manifest_view);
+
+            if (!ntp_config.is_read_replica_mode_enabled()) {
+                _upload_housekeeping.register_jobs(
+                  archiver->get_housekeeping_jobs());
+            }
+            return archiver;
+        }
+        return nullptr;
+    }
+
+    /// Create and start archiver. Return started archiver instance.
+    ss::future<ss::lw_shared_ptr<ntp_archiver>>
+    bg_create_and_start_archiver(model::term_id expected_term) noexcept {
+        using ret_t = ss::lw_shared_ptr<ntp_archiver>;
+        try {
+            vlog(
+              _ctxlog.debug,
+              "managed_partition: starting ntp_archiver in term {}",
+              expected_term);
+
+            auto archiver = maybe_construct_archiver();
+
+            if (archiver != nullptr) {
+                return archiver->start().then_wrapped(
```

Comment:
what happens if the term changed by the time we got there? and to the later log statements